### PR TITLE
Port Chemistry Checker grammar to Nearley.js

### PR DESCRIPTION
**Pull request by skyepurchase**
As part of the move to an improved Chemistry Checker the grammar that was defined in [the Chemistry Checker](https://github.com/isaacphysics/chemistry-checker) has been ported to Nearley.js. This is placed in Inequality Grammar to keep all of the lexers and parsers in one location.

There are some differences between the original grammar and the ported grammar:
* The new grammar is split into Chemistry and Nuclear Physics to simplify the grammar and processing.
* The grammars have been simplified as they do not need to deal with both grammars at once.
    * There have also been some minor improvements to the grammar rules.
* Additional tokens and structures are now correctly parsed as these were found in the current question set.
    * The previous grammar relied on string match shortcutting resulting in overly liberal correct answers in the question.

Unit tests have been added to Inequality Grammar! There are only tests for the new lexers and parsers. This was done to make reviewing the code easier given the size of the grammars and deviation from the original grammar. The tests also allow for new features to be properly defined and implemented rather than adhoc implementation.

The repository experiences very little traffic and so there is little need for CI. Similarly, the existing grammars have remained stable for extended periods of time.

### CODE REVIEW

File: assets/chemistry-grammar.ne [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    })},
+
+    // Subscripts
+    Sub: /_{[1-9][0-9]*}/,
+
+    // Non-zero naturals
+    Num: /[1-9][0-9]*/,
+
+    // Fractions
+    Frac: /\\frac{[1-9][0-9]*}{[1-9[0-9]*}/,
```
Comment (jsharkey13) [inline assets/chemistry-grammar.ne:30]: This won't cope with `1/2O2`, which is a perfectly acceptable way of writing fractions in mhchem.

File: assets/nuclear-grammar.ne [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    Charge: { match: /(?:-|\^{(?:[1-9][0-9]*)?(?:\+|\-)})/, type: moo.keywords({
+        Positive: "^{+}",
+        Negative: ["^{-}", "-"]
+    })},
+
+    // Non-zero naturals
+    Num: /[1-9][0-9]*/,
+
+    // Fractions
+    Frac: /\\frac{[1-9][0-9]*}{[1-9[0-9]*}/,
```
Comment (jsharkey13) [inline assets/nuclear-grammar.ne]: Do we need fractions for nuclear stuff? Fractions are common in chemistry for standard enthalpy equations (where what is important is one unit of a compound burning, which may require a fractional quantity of oxygen and create fractional products), but I don't see fractional atoms likely in nuclear physics!

File: package.json [hunk 3]
```diff
@@ -21,14 +22,18 @@
...
+    "babel-jest": "^29.7.0",
     "babel-loader": "^8.2.5",
+    "jest": "^29.7.0",
     "nearley-loader": "^2.0.0",
     "webpack": "^5.74.0",
     "webpack-cli": "^4.10.0"
   },
   "dependencies": {
+    "inequality-grammar": "^1.1.0",
+    "jest-transform-nearley": "^2.0.0",
```
Comment (jsharkey13) [inline package.json]: Could this not be a dev dependency?

File: assets/chemistry-grammar.ne (hunk 1 above)
Comment (skyepurchase) [reply assets/chemistry-grammar.ne:30]: Will add this case in. Will take any pattern `<num>/<num>` as a fraction so `1/202` is never `1/20 2` or similar

File: assets/nuclear-grammar.ne (hunk 2 above)
Comment (skyepurchase) [reply assets/nuclear-grammar.ne]: Will happily remove this functionality!

File: package.json (hunk 3 above)
Comment (skyepurchase) [reply package.json]: The `jest-transform-nearley` was meant to be a dev dependency. Really concerned about it depending on itself, will definitely remove that 😅.

**Review by jsharkey13 — APPROVED** [review]

**Pull request merged**